File: app/src/main/java/com/wmods/wppenhacer/xposed/core/components/AlertDialogWpp.java
package com.wmods.wppenhacer.xposed.core.components;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.view.View;
import android.widget.Toast;

import com.wmods.wppenhacer.xposed.core.devkit.Unobfuscator;
import com.wmods.wppenhacer.xposed.utils.ReflectionUtils;
import com.wmods.wppenhacer.xposed.utils.Utils;

import java.lang.reflect.Method;

import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;

public class AlertDialogWpp {


    private static Method getAlertDialog;
    private static Method setItemsMethod;
    private static boolean isAvailable;
    private static Method setMessageMethod;
    private static Method setNegativeButtonMethod;
    private static Method setPositiveButtonMethod;
    private static Method setMultiChoiceItemsMethod;
    private final Context mContext;
    private AlertDialog.Builder mAlertDialog;
    private Object mAlertDialogWpp;
    private Dialog mCreate;

    public static void initDialog(ClassLoader loader) {
        try {
            getAlertDialog = Unobfuscator.loadMaterialAlertDialog(loader);
            Class<?> alertDialogClass = getAlertDialog.getReturnType();
            setItemsMethod = ReflectionUtils.findMethodUsingFilter(alertDialogClass, method -> method.getParameterCount() == 2 && method.getParameterTypes()[0].equals(DialogInterface.OnClickListener.class) && method.getParameterTypes()[1].equals(CharSequence[].class));
            setMultiChoiceItemsMethod = ReflectionUtils.findMethodUsingFilter(alertDialogClass, method -> method.getParameterCount() == 3 && method.getParameterTypes()[0].equals(DialogInterface.OnMultiChoiceClickListener.class) && method.getParameterTypes()[1].equals(CharSequence[].class));
            setMessageMethod = ReflectionUtils.findMethodUsingFilter(alertDialogClass, method -> method.getParameterCount() == 1 && method.getParameterTypes()[0].equals(CharSequence.class));
            var buttons = ReflectionUtils.findAllMethodsUsingFilter(alertDialogClass, method -> method.getParameterCount() == 2 && method.getParameterTypes()[0].equals(DialogInterface.OnClickListener.class) && method.getParameterTypes()[1].equals(CharSequence.class));
            setNegativeButtonMethod = buttons[0];
            setPositiveButtonMethod = buttons[2];
            isAvailable = true;
        } catch (Throwable e) {
            isAvailable = false;
            XposedBridge.log(e);
            Utils.showToast("Failed to load MaterialAlertDialog", Toast.LENGTH_SHORT);
        }
    }

    public AlertDialogWpp(Context context) {
        mContext = context;
        if (isSystemDialog()) {
            mAlertDialog = new AlertDialog.Builder(context);
            return;
        }
        try {
            mAlertDialogWpp = getAlertDialog.invoke(null, context);
            // Remove Default Message0
            setMessage(null);
        } catch (Exception ignored) {
            throw new RuntimeException("Failed to create AlertDialogWpp");
        }
    }

    public Context getContext() {
        return mContext;
    }

    public static boolean isSystemDialog() {
        return !isAvailable;
    }

    public AlertDialogWpp setTitle(String title) {
        if (isSystemDialog()) {
            mAlertDialog.setTitle(title);
            return this;
        }
        XposedHelpers.callMethod(mAlertDialogWpp, "setTitle", title);
        return this;
    }

    public AlertDialogWpp setTitle(int title) {
        if (isSystemDialog()) {
            mAlertDialog.setTitle(title);
            return this;
        }
        XposedHelpers.callMethod(mAlertDialogWpp, "setTitle", getContext().getString(title));
        return this;
    }

    public AlertDialogWpp setMessage(String message) {
        if (isSystemDialog()) {
            mAlertDialog.setMessage(message);
            return this;
        }
        try {
            setMessageMethod.invoke(mAlertDialogWpp, message);
        } catch (Exception ignored) {
        }
        return this;
    }

    public AlertDialogWpp setItems(CharSequence[] items, DialogInterface.OnClickListener listener) {
        if (isSystemDialog()) {
            mAlertDialog.setItems(items, listener);
            return this;
        }
        try {
            setItemsMethod.invoke(mAlertDialogWpp, listener, items);
        } catch (Exception e) {
            XposedBridge.log(e);
        }
        return this;
    }


    public AlertDialogWpp setMultiChoiceItems(CharSequence[] items, boolean[] checkedItems, DialogInterface.OnMultiChoiceClickListener listener) {
        if (isSystemDialog()) {
            mAlertDialog.setMultiChoiceItems(items, checkedItems, listener);
            return this;
        }
        try {
            setMultiChoiceItemsMethod.invoke(mAlertDialogWpp, listener, items, checkedItems);
        } catch (Exception ignored) {
        }
        return this;
    }

    public AlertDialogWpp setNegativeButton(CharSequence text, DialogInterface.OnClickListener listener) {
        if (isSystemDialog()) {
            mAlertDialog.setNegativeButton(text, listener);
            return this;
        }
        try {
            setNegativeButtonMethod.invoke(mAlertDialogWpp, listener, text);
        } catch (Exception ignored) {
        }
        return this;
    }

    public AlertDialogWpp setPositiveButton(CharSequence text, DialogInterface.OnClickListener listener) {
        if (isSystemDialog()) {
            mAlertDialog.setPositiveButton(text, listener);
            return this;
        }
        try {
            setPositiveButtonMethod.invoke(mAlertDialogWpp, listener, text);
        } catch (Exception ignored) {
        }
        return this;
    }

    public AlertDialogWpp setView(View view) {
        if (isSystemDialog()) {
            mAlertDialog.setView(view);
            return this;
        }
        XposedHelpers.callMethod(mAlertDialogWpp, "setView", view);
        return this;
    }


    public Dialog create() {
        if (mCreate != null) return mCreate;
        if (isSystemDialog()) {
            mCreate = mAlertDialog.create();
        } else {
            mCreate = (Dialog) XposedHelpers.callMethod(mAlertDialogWpp, "create");
        }
        return mCreate;
    }

    public void dismiss() {
        if (mCreate == null) return;
        mCreate.dismiss();
    }

    public void show() {
        if (isSystemDialog()) {
            mAlertDialog.show();
            return;
        }
        create().show();
    }

}

File: app/src/main/java/com/wmods/wppenhacer/xposed/core/components/AlertDialogWpp.kt
package com.wmods.wppenhacer.xposed.core.components

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.DialogInterface.OnMultiChoiceClickListener
import android.view.View
import android.widget.Toast
import com.wmods.wppenhacer.xposed.core.devkit.Unobfuscator.loadMaterialAlertDialog
import com.wmods.wppenhacer.xposed.utils.ReflectionUtils
import com.wmods.wppenhacer.xposed.utils.Utils
import de.robv.android.xposed.XposedBridge
import de.robv.android.xposed.XposedHelpers
import java.lang.reflect.Method

open class AlertDialogWpp(val context: Context?) {
    private var mAlertDialog: AlertDialog.Builder? = null
    private var mAlertDialogWpp: Any? = null
    private var mCreate: Dialog? = null

    init {
        if (isSystemDialog) {
            mAlertDialog = AlertDialog.Builder(context)
        }else {
            try {
                mAlertDialogWpp = getAlertDialog!!.invoke(null, context)
                setMessage(null)
            } catch (_: Exception) {
                throw RuntimeException("Failed to create AlertDialogWpp")
            }
        }
    }

    fun setTitle(title: String?): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setTitle(title)
            return this
        }
        XposedHelpers.callMethod(mAlertDialogWpp, "setTitle", title)
        return this
    }

    fun setTitle(title: Int): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setTitle(title)
            return this
        }
        XposedHelpers.callMethod(mAlertDialogWpp, "setTitle", this.context!!.getString(title))
        return this
    }

    fun setMessage(message: CharSequence?): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setMessage(message)
            return this
        }
        try {
            setMessageMethod!!.invoke(mAlertDialogWpp, message)
        } catch (_: Exception) {
        }
        return this
    }

    fun setItems(
        items: Array<CharSequence?>?,
        listener: DialogInterface.OnClickListener?
    ): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setItems(items, listener)
            return this
        }
        try {
            setItemsMethod!!.invoke(mAlertDialogWpp, listener, items)
        } catch (e: Exception) {
            XposedBridge.log(e)
        }
        return this
    }


    fun setMultiChoiceItems(
        items: Array<String>,
        checkedItems: BooleanArray?,
        listener: OnMultiChoiceClickListener?
    ): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setMultiChoiceItems(items, checkedItems, listener)
            return this
        }
        try {
            setMultiChoiceItemsMethod!!.invoke(mAlertDialogWpp, listener, items, checkedItems)
        } catch (_: Exception) {
        }
        return this
    }

    fun setNegativeButton(
        text: CharSequence?,
        listener: DialogInterface.OnClickListener?
    ): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setNegativeButton(text, listener)
            return this
        }
        try {
            setNegativeButtonMethod!!.invoke(mAlertDialogWpp, listener, text)
        } catch (_: Exception) {
        }
        return this
    }

    fun setPositiveButton(
        text: CharSequence?,
        listener: DialogInterface.OnClickListener?
    ): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setPositiveButton(text, listener)
            return this
        }
        try {
            setPositiveButtonMethod!!.invoke(mAlertDialogWpp, listener, text)
        } catch (_: Exception) {
        }
        return this
    }

    fun setView(view: View?): AlertDialogWpp {
        if (isSystemDialog) {
            mAlertDialog!!.setView(view)
            return this
        }
        XposedHelpers.callMethod(mAlertDialogWpp, "setView", view)
        return this
    }


    open fun create(): Dialog {
        if (mCreate != null) return mCreate!!
        mCreate = if (isSystemDialog) {
            mAlertDialog!!.create()
        } else {
            XposedHelpers.callMethod(mAlertDialogWpp, "create") as Dialog
        }
        return mCreate!!
    }

    fun dismiss() {
        if (mCreate == null) return
        mCreate!!.dismiss()
    }

    fun show() {
        if (this.context is Activity) {
            val activity = this.context
            if (activity.isFinishing || activity.isDestroyed) {
                return
            }
        }
        if (isSystemDialog) {
            mAlertDialog!!.show()
            return
        }
        create()!!.show()
    }

    companion object {
        private var getAlertDialog: Method? = null
        private var setItemsMethod: Method? = null
        private var isAvailable = false
        private var setMessageMethod: Method? = null
        private var setNegativeButtonMethod: Method? = null
        private var setPositiveButtonMethod: Method? = null
        private var setMultiChoiceItemsMethod: Method? = null
        fun initDialog(loader: ClassLoader) {
            try {
                getAlertDialog = loadMaterialAlertDialog(loader)
                val alertDialogClass: Class<*> = getAlertDialog!!.returnType
                setItemsMethod = ReflectionUtils.findMethodUsingFilter(
                    alertDialogClass
                ) { method: Method? -> method!!.parameterCount == 2 && method.parameterTypes[0] == DialogInterface.OnClickListener::class.java && method.parameterTypes[1] == Array<CharSequence>::class.java }
                setMultiChoiceItemsMethod = ReflectionUtils.findMethodUsingFilter(
                    alertDialogClass
                ) { method: Method? -> method!!.parameterCount == 3 && method.parameterTypes[0] == OnMultiChoiceClickListener::class.java && method.parameterTypes[1] == Array<CharSequence>::class.java }
                setMessageMethod = ReflectionUtils.findMethodUsingFilter(
                    alertDialogClass
                ) { method: Method? -> method!!.parameterCount == 1 && method.parameterTypes[0] == CharSequence::class.java }
                val buttons = ReflectionUtils.findAllMethodsUsingFilter(
                    alertDialogClass
                ) { method: Method? -> method!!.parameterCount == 2 && method.parameterTypes[0] == DialogInterface.OnClickListener::class.java && method.parameterTypes[1] == CharSequence::class.java }
                setNegativeButtonMethod = buttons[0]
                setPositiveButtonMethod = buttons[2]
                isAvailable = true
            } catch (e: Throwable) {
                isAvailable = false
                XposedBridge.log(e)
                Utils.showToast("Failed to load MaterialAlertDialog", Toast.LENGTH_SHORT)
            }
        }

        val isSystemDialog: Boolean
            get() = !isAvailable
    }
}
